### PR TITLE
Add rotation to plants.

### PR DESCRIPTION
**Pull request by RealBadAngel**

### CODE REVIEW

**Comment by paramat** [comment]
I feel random rotation is unsuitable for any plant during mapgen or when farming, although the ability to rotate the plantlike drawtype is good for special uses by modders and i have nothing against that.

The way that plantlike textures are at 45 degrees to the world grid is very much part of the character of a voxel world, randomising the rotation is essentially like randomising the rotation of voxels (if that were possible) which would obviously be ridiculous.

Also, the smoothness and realism of random rotation is unsuitable in character for the extreme simplicity and abstraction of the plantlike drawtype, which needs a suitably simple and abstract rotation system.

When grasses are dense in mgv5/v7 they would look bad randomly rotated, just the same way a field of wheat looks wrong with random rotation. In mgv6 flowers and grasses are spaced far apart so the rotation doesn't look so bad.

**Comment by HybridDog** [comment]
>I feel random rotation is unsuitable for any plant during mapgen or when farming, and just looks wrong in a voxel world, although the ability to rotate the plantlike drawtype is good for special uses by modders and i have nothing against that.

l think wheat should not be rotated randomly, but 45°, because the player puts the seeds on straight tracks he made with a hoe in soil.
But plants like weed should be rotated randomly to make it look more messy.

**Comment by paramat** [comment]
As deputy mapgen dev i am very strongly opposed to this. Mgv5/mgv7 now have higher density grass, the way the 45 degree orientations line up creates an aligned, patterned, regular look that is as important to Minetest as aligned voxels are, no one would ever randomly rotate voxels in a voxel world.
Flowers are so small the random rotation is not very noticeable anyway.

**Comment by kilbith** [comment]
@paramat It's exactly like intend to make all trees +- similar, undiversified and boring... just opposed to how vegetal life inherently is.

**Comment by HybridDog** [comment]
maybe that's the reason why new trees hardly become added to minetest although [its on the todo list](http://dev.minetest.net/todo#Add_things_to_surface)

**Comment by OttoLidenbrock** [comment]
@paramat 
>The way that plantlike textures are at 45 degrees to the world grid is very much part of the character of a voxel world

There is no such thing as "the character of a voxel world". Please stop imposing your views (i.e. things must always be square and aligned) on others.

>no one would ever randomly rotate voxels in a voxel world.

Except people who want to make things look good and not be stuck with some per-concieved conception of what a voxel world must look like.
 
> This is so out of character it should not even be considered as an option.

Out of character? See above.

**Comment by paramat** [comment]
>  all trees +- similar, undiversified and boring... just opposed to how vegetal life inherently is.

Minetest is not meant to be realistic, a voxel world is inherently aligned, simple and abstract.

> There is no such thing as "the character of a voxel world"

Obviously there is, 1m cubes orientated to a world grid, plantlike textures in a cross diagonally orientated.
I'm just expressing my strong feelings, it's impossible to impose my views on others, people have their own views. Excessive moves towards realism, photo-realism, messiness, smoothness and high-resolution are so obviously a mistake in a voxel game, i'm just pointing out the blindingly obvious in exasperation.

> Except people who want to make things look good and not be stuck with some per-concieved conception of what a voxel world must look like.

Randomly rotated voxels are impossible in Minetest, it would not be a voxel game if you did. Does anyone really want to change Minetest into a photrealistic game? if so you should really work on and play another game.

**Comment by kilbith** [comment]
> Minetest is not meant to be realistic

Why not making the grass purple or the water green then, if it's "inherently" nicer ?

> a voxel world is inherently aligned, simple and abstract.

BTW, there are some things not aligned in Minecraft (eg. the banners).

> i'm just pointing out the blindingly obvious in exasperation that many players just don't understand or value the character of Minetest.

You have your conception of a voxel game, good -- but simply your very formal way of defining/decreeing what should be Minetest may be exasperating for some of us. This restricts the immersivity and diversity.

Beyond that you're making a good work for the biomes, overall.

**Comment by C1ffisme** [comment]
I for one say that @paramat has a point. We can't move to photoreality in a voxel game. We may as well just remove digging and building and turn this into a house decor game.

If there is anything that dosen't belong in a voxel game, it's "over-realism". Just try playing with Vannesa's 1024px textures. The game dosen't look real, it looks terrible. (No offense meant, Vannessa.) See, voxel games are meant to have cute textures that imitate reality, but don't exactly copy it.

On the other hand, adding rotation to plants isn't that far of a problem. I mean, plants should be aligned with plant collision boxes anyway. So we already have diagonals, why can't plants just have a couple extra?

**Comment by paramat** [comment]
The way i see it, the plantlike drawtype is so extremely simple it looks wrong with anything other than an extremely simple diagonal orientation, the screenshots of randomly rotated wheatfields look really bad, a mess that looks out of place in an iherently aligned voxel world, this is what dense grassland in mgv5/v7 would look like.

If you are going as far as random rotation it's better to not use plantlike drawtype but instead 3D model plants in micro-voxels. But then ironically enough, you would only have 4 rotations available. If only 4 rotations are possible with 3D models there is no need more than 4 for plantlike.

Plantlike rotation is available and i actually support it for specialised use in mods and subgames, i'm sure there are some very cool uses. Randomly rotated plants are fine in subgames because subgames are more specialised, but please keep it out of basic mapgen and our basic subgame MTGame.

**Comment by paramat** [comment]
The problem with this issue is i feel too strongly about it =) and am beginning to repeat myself, so am deleting some of my posts.

**Comment by kilbith** [comment]
@paramat This totally breaks the consistency of the discussion...

**Comment by Zeno-** [comment]
Well, they are right.

The ID tech 6 (Doom 4) engine was originally planned to be voxel based. I'm pretty sure that Carmack and ID Software didn't propose developing an engine that only supported pixel art, limited rotation, no advanced lighting, etc, etc. So, I'd have to agree that there is no such thing as a voxel world in the context that it's being used in this discussion. Perhaps you're all talking about a minecraft world when you're trying to define the character of a voxel world?

**Comment by paramat** [comment]
I was repeating myself and being unpleasant, i don't want that sort of negativity hanging around on the internet =)
Zeno, your examples are of games evolving into something else, i feel Minetest should remain a fairly pure voxel world. Such a world has grid-aligned nodes with aligned rotations, this includes all drawtypes. Although it does not seem like it (because plantlike nodes are not very common in mgv6) fine-step random rotations of plantlike is as radical a change as fine-step random rotation of voxels, which (if it were possible) would of course be completely out of the question. Finely rotated plantlike nodes look bad and out-of-place in a regular aligned, simple and abstract voxel world.
There are some Minetest features that are not used extensively in basic mapgen for good reason, because they are more specialised features for subgames and mods.

**Comment by paramat** [comment]
There are many photorealistic games available, but Minetest is perhaps the only excellent voxel game alternative to Minecraft, so i feel its simple abstraction should be preserved and not evolve towards more realism. The whole idea of a voxel game is simplicity, abstraction and lack of fancy graphics to focus resources on more interesting things like having a real-time rendered highly modifiable world.

A mixture of voxel world and fancy realism just looks very wrong, see Creativerse as an example, it is hideous https://www.youtube.com/watch?v=gELgJDDMZBo The realistic and fancy graphic elements make the voxel structure look ridiculous and the voxel structure make the realism look ridiculous, this is the problem with a halfway design style. This in general is what i am resisting in my Github comments.

**Comment by C1ffisme** [comment]
You know, I feel we're all overreacting. Let's break it down to some questions.

1. Does this affect gameplay?
Yes, it adds to it, in fact. Players want their wild grass to look, well, wild? Yes.
2. Does this make minetest look cheap?
If someone were to use a screwdriver on their wheat fields to make them look random and wierd and took a screenshot to show to everybody, then they might accomplish this. Why they would do this? No idea. (Maybe a guy who dosen't like Minecraft clones?) On the other hand, this dosen't have to be added to wheat. I'd rather wheat stayed proportioned, while wild grass stayed wild.
3. Does this make minetest look too photorealistic?
No. Real plants don't form an X. If we were going for photorealistic, we'd make all our plants into meshnodes.
4. Is this worth merging?
Yes. I think so. Like paramat said, we should still go for 'uncanny valley', but I doubt this is a change as significant as animating the elbows and knees of the player.

**Comment by paramat** [comment]
If you ask celeron55 if realism is a purpose or goal of Minetest he will say 'not in any way'. The essential character is K.I.S.S. This feature is fine for specialised subgames and mods but not basic default mapgen.

Aside from the design discussion, the biome API cannot set param2 for simple decorations, the 'simple decoration' code is meant to be as simple and light as possible because it is used so much during mapgen. You would need to convince hmmmm to approve the code being made more complex for something completely out of character that is applied to long grasses only. There is no point adding this to flowers as they are so small it would be almost unnoticeable unless you are up close.

**Comment by PilzAdam** [comment]
@C1ffisme I don't see how visual changes like this can affect gameplay. Can you elaborate on this?

I'm with @paramat on this. Single nodes (blocks or plants) should all "look the same". The "randomness" or "wildness" is not achieved by randomizing single nodes, but rather by the placement via the mapgen.

**Comment by Zeno-** [comment]
@PilzAdam Why should they look the same? How does some kind of random rotation detract from the "character" of a voxel game? "Wildness" most certainly *is* achieved my rotating single nodes IMO

**Comment by Zeno-** [comment]
@paramat 

> You would need to convince hmmmm to approve the code being made more complex for something completely out of character that is applied to long grasses only

I don't believe you. Sorry.

**Comment by HybridDog** [comment]
>A mixture of voxel world and fancy realism just looks very wrong, see Creativerse as an example, it is hideous https://www.youtube.com/watch?v=gELgJDDMZBo The realistic and fancy graphic elements make the voxel structure look ridiculous and the voxel structure make the realism look ridiculous, this is the problem with a halfway design style when one or the other has a design consistency. This in general is what i am resisting in my Github comments.

What about this?
https://www.youtube.com/watch?list=PL-btRGIvRn5ncTG0X_lIazJLkWORZQQ0O&v=5fACj-SOzmY

**Comment by Zeno-** [comment]
@HybridDog Fuck. You win. Thread closed.

**Comment by rubenwardy** [comment]
Creativerse isn't that bad, the lava looks out of place. I'd say it's more cartoony than realistic. BuildAWorld is what i wish minetest was, it's heavily physics based and has cool gameplay mechanisms. But that's offtopic, graphic wise it is very good.

You can have good graphics in a sandbox game. I have yet to try this patch out, and I can't see any screen shots. I'd say both placement and rotation play a part in the wildness of plants etc.

File: mods/default/mapgen.lua [hunk 1]
```diff
@@ -324,12 +324,13 @@ function default.generate_ore(name, wherein, minp, maxp, seed, chunks_per_volume
 end
 
 function default.make_papyrus(pos, size)
+	local rot = math.random(0,179)
```
Comment (rubenwardy) [inline mods/default/mapgen.lua]: Space between parameters please :P

**Comment by paramat** [comment]
I'm still very strongly against this. It would also need an engine change as these decorations are now placed by the biome API decoration code, which currently cannot set param2 because the code is meant to be simple.
Having a problem with a diagonal grid of plants is like having a problem with cubic nodes aligned to the world grid, this is a ridiculous opinion to have in a voxel game, best i think this is kept to RBA's fork.

**Comment by 0-afflatus** [comment]
This is an engine issue for sure. 

While I applaud efforts to keep mapgen simple, I find it counter-intuitive that the 'rotate' parameter only rotates schematics and not nodes. Decoration can be used for a multitude of things, not just grass placement and I would like to be able to randomly place directionally oriented nodes, but I think only four directions need supporting 0, 90, 180 and 270 degrees.

AFA Minetest Game is concerned I would vote :-1: but my opinion isn't necessarily relevant.

**Comment by paramat** [comment]
>  I would like to be able to randomly place directionally oriented nodes, but I think only four directions need supporting 0, 90, 180 and 270 degrees.

Yes i agree now to supporting 4 rotations for simple decorations, waterlilies would need them for example. I'll probably add this feature soon. However not the fine-step rotation of plants that have paramtype2 = degrotate.

**Comment by paramat** [comment]
Closing because no replies from the author for a year, also papyrus, dry shrub, long grasses and flowers are now all placed by the decoration API which does not currently set param2, so this PR is unusable and this is now an engine issue.

**Pull request closed without merging**